### PR TITLE
Refactor SYSV init, initramfs and dracut (Phase 3)

### PR DESCRIPTION
**Pull request by FransUrbo**
- Move code from the dracut scripts that is better served in a common code base library (zfs-functions).
- Add functionality that exists in initramfs to dracut.
  - Support 'zfs:<rootfs>' import and mounting. This was for some reason removed in eda3d4e1de94cd3a9247bb90a64413a89e736847.
  - Support booting from snapshots.
  - Support mounting recursive filesystems.
  - Support mounting of additional filesystems.
  - Support mounting nativly encrypted filesystems.

This depend on #3559, this PR is "Phase 3". It seriously modifies the dracut code (tested extensively on Fedora 17).

Because #3559 have not been accepted yet, this PR includes those commits as well. But once that PR have been accepted (eventually?), I do a new push and they should vanish.

Even though the intention is to get dracut accepted in dracut upstream, there is no reason why not to do this. The <code>zfs-functions</code> could either be pushed there "as is" or code used could be extracted. Considering the added functionality, which would be next to impossible (and extremely difficult to maintain), giving ALL systems the exact same boot parameters and functionality is, in my opinion, extremely important for a project as ZoL.

Therefor, this should then be tagged 0.6.6 as earliest, but I'm good with 0.6.7. I DO think it needs to be done eventually though.

### CODE REVIEW

**Comment by Aviator45003** [comment]
It would be convenient if this were merged at some point. I am having troubles with my `/usr` being on a different zfs filesystem.

**Comment by FransUrbo** [comment]
You know, you CAN merge it yourself :). That would mean more people testing it and verifying that it works, the sooner it will be merged.

**Comment by FransUrbo** [comment]
@behlendorf Do we have an (approximate) ETA on this?

**Comment by behlendorf** [comment]
I'd like to get more eyes on these changes.  @ryao can you look these over.

**Comment by Vaelatern** [comment]
Is there a method better than manually merging diffs to merge this to a running system?

**Comment by Vaelatern** [comment]
I applied the patch, rebuilt zfsonlinux on my distro of choice (Void Linux). All the files seem to be in the right place, but reconfiguring the linux kernel (which on this setup triggers a dracut rebuild) I got `dracut-install: ERROR: installing '@initconfdir@/zfs'` followed by `dracut: /usr/lib/dracut/dracut-install -D /tmp/dracut.QsJXN6/initramfs @initconfdir@/zfs`.
I can't figure out how to further test the line. If I manually run the aforementioned command with `s:@initconfdir@/:/usr/lib/dracut/modules.d/90` then it completes with no further errors (but have not tested if it boots).

File: contrib/dracut/90zfs/module-setup.sh.in [hunk 1]
```diff
@@ -39,11 +41,15 @@ install() {
...
 	dracut_install head
+	dracut_install sort
+	dracut_install grep
 	inst_hook cmdline 95 "${moddir}/parse-zfs.sh"
 	inst_hook mount 98 "${moddir}/mount-zfs.sh"
 	inst_hook shutdown 30 "${moddir}/export-zfs.sh"
 
-	inst_simple "${moddir}/zfs-lib.sh" "/lib/dracut-zfs-lib.sh"
+	inst_simple "@sysconfdir@/zfs/zfs-functions" "/lib/dracut-zfs-lib.sh"
+	inst "@initconfdir@/zfs"
```
Comment (Vaelatern) [inline contrib/dracut/90zfs/module-setup.sh.in]: Is this truly desired? It is not obvious what it replaces.

Comment (Vaelatern) [inline contrib/dracut/90zfs/module-setup.sh.in]: Bug. This file is not parsed by `make` to parse out `initconfdir`. It parses out `sysconfdig` but not `initconfdir`.

**Comment by FransUrbo** [comment]
@Vaelatern With your fix (which I've just merged and force-pushed), does it work as expected?

Do you want to add your **Signed-off-by**?

**Comment by Vaelatern** [comment]
The patch now looks right. Yes, it works on one of my machines. I'm happy to add my **Signed-off-by**

File: contrib/dracut/90zfs/module-setup.sh.in (hunk 1 above)
Comment (Vaelatern) [inline contrib/dracut/90zfs/module-setup.sh.in]: Bug resolved with patch and rebase.

**Comment by FransUrbo** [comment]
No, I merged it to the first commit (Refactoring of common code between the SYSV init, initramfs and dracut scripts.). It made no sense having it in it's separate commit.

**Comment by Vaelatern** [comment]
I was referring to my Signed-off-by, I'm not sure if I add it or if you.

**Comment by FransUrbo** [comment]
Ah, that one. No, that have not been added yet. It's in here, if @behlendorf ever feels like this might be important or relevant to look at.

**Comment by ryao** [comment]
Why is this in so many different commits?

**Comment by prometheanfire** [comment]
someone mind sharing their kernel command line? `root=zfs:AUTO` didn't work.

**Comment by prometheanfire** [comment]
here's the output, it also didn't try to import the pool https://gist.github.com/prometheanfire/5dc33fcc55579bc6fd56

**Comment by Vaelatern** [comment]
@prometheanfire, is "AUTO" your pool name?

**Comment by prometheanfire** [comment]
nope, I'll try changing that then :D

**Comment by prometheanfire** [comment]
and it didn't work with zfs:mthode-work (name of the pool)

the drive is decrypted then the mount is called and fails.  within the recovery shell I can import the pool and all still.

**Comment by prometheanfire** [comment]
it seems to me that the mount command doesn't understand zfs:foo or something...

**Comment by Vaelatern** [comment]
The feature is specifically `Support 'zfs:<rootfs>' import and mounting` so try that?

**Comment by prometheanfire** [comment]
zfs:mthode-work/OS did not work

**Comment by prometheanfire** [comment]
here's a portion of the rdsoreport.txt


    [   13.006331] localhost systemd[1]: Mounting /sysroot...
    [   13.006919] localhost mount[3806]: mount: wrong fs type, bad option, bad superblock on zfs:mthode-work/OS,
    [   13.007006] localhost mount[3806]: missing codepage or helper program, or other error
    [   13.007088] localhost mount[3806]: (for several filesystems (e.g. nfs, cifs) you might
    [   13.007161] localhost mount[3806]: need a /sbin/mount.<type> helper program)
    [   13.007234] localhost mount[3806]: In some cases useful info is found in syslog - try
    [   13.007317] localhost mount[3806]: dmesg | tail or so.

and the systemctl status

    ● sysroot.mount - /sysroot
       Loaded: loaded (/proc/cmdline; bad; vendor preset: enabled)
       Active: active (mounted) (Result: exit-code) since Sun 2016-02-07 05:20:36 UTC; 24s ago
        Where: /sysroot
         What: /dev/sda1
         Docs: man:fstab(5)
               man:systemd-fstab-generator(8)
      Process: 3806 ExecMount=/bin/mount zfs:mthode-work/OS /sysroot -o ro (code=exited, status=32)

**Comment by prometheanfire** [comment]
seems to be trying `mount zfs:mthode-work/OS` instead of `mount -t zfs`, not sure if that's intentional.

**Comment by prometheanfire** [comment]
tried manual mounting in the initram with mount -t zfs, failed because couldn't find the dataset, imported dataset with -N and tried again, failed because dataset was not marked as legacymount, which I don't think I should change.

**Comment by Vaelatern** [comment]
@prometheanfire, are you sure you are using zfs with the above patches actually applied? At no point should it be trying to mount "zfs:rpool" but rather importing "rpool"

**Comment by prometheanfire** [comment]
I'm sure...

here is /usr/lib64/dracut/modules.d/90zfs/mount-zfs.sh https://bpaste.net/show/e3067d32ae37

**Comment by FransUrbo** [comment]
@ryao it includes the commits from "Phase 2" as well. Only the last three is relevant to the Dracut changes.

@prometheanfire <code>root=zfs:AUTO</code> **is** suppose to work! But on the other hand, I've never tested this using <code>systemd</code>!!

**Comment by prometheanfire** [comment]
@FransUrbo if you have time on I'm on the freenode channel as prometheanfire, we can debug if you want.

**Comment by Vaelatern** [comment]
Where is the status of this? Were the systemd issues worked out yet?

**Comment by Vaelatern** [comment]
According to the other thread, the systemd issues are elsewhere, and not related to this patch, #3605.

Is this likely to be merged or shall I adopt the last three commits, rewrite them to not require any of the preceding commits, and PR that (with appropiate credit given)?

File: config/zfs-build.m4 [hunk 2]
```diff
@@ -334,6 +334,16 @@ AC_DEFUN([ZFS_AC_DEFAULT_PACKAGE], [
 		AC_MSG_RESULT([no])
 	fi
 	AC_SUBST(DEFINE_INITRAMFS)
+
+	AC_MSG_CHECKING([whether dracut is available])
+	if test -d /usr/lib/dracut ; then
+		DEFINE_DRACUT='--define "_dracut 1"'
```
Comment (Rudd-O) [inline config/zfs-build.m4]: 1. Is `/usr/share/dracut` no longer supported?
2. Is the check to be performed by the host building the package?  Doesn't this break building an F23 package in an F22 host, or building the package in a host that has no Dracut?  Or are we all already deep into the Mock ideology that this is a lost cause?

Just wondering.

File: contrib/Makefile.am [hunk 3]
```diff
@@ -1,2 +1,2 @@
-SUBDIRS = bash_completion.d dracut initramfs
-DIST_SUBDIRS = bash_completion.d dracut initramfs
+SUBDIRS = bash_completion.d dracut initramfs sysv-init shell-common
+DIST_SUBDIRS = bash_completion.d dracut initramfs sysv-init shell-common
```
Comment (Rudd-O) [inline contrib/Makefile.am]: In the vernacular, we call this "sum gud shit".

It's looking good, man.

Comment (FransUrbo) [inline contrib/Makefile.am]: Oups, didn't notice that! :)

**Comment by FransUrbo** [comment]
If (when!?) this actually attracts some actual interest to be merged, I can spend time on it. But not a second until then!

I've been tricked before.

**Comment by behlendorf** [comment]
There are some nice improvements in these proposed patches.  Unfortunately, they never got the attention they deserved in terms of reviewers and testing.  As such I was never comfortable with merging them to master.  If someone would like to continue with this work my suggestion would be to start with the most important changes and open new PRs against master.  Keeping the PRs small and as functionally independent as possible will make it easier to review and test the changes.  Closing.

**Pull request closed without merging**